torch/csrc/jit/script/tree_views.h

```diff
@@ -20,423 +91,422 @@ struct TreeView {
   }
-  typedef ListViewIterator<T> iterator;
-  typedef ListViewIterator<T> const_iterator;
   iterator begin() const {
     return iterator(tree_->trees().begin());
   }
   iterator end() const {
     return iterator(tree_->trees().end());
   }
   T operator[](size_t i) const {
-    return T(tree_->trees().at(i));
+    return T(subtree(i));
   }
   TreeRef map(std::function<TreeRef(const T&)> fn) {
     return tree_->map([&](TreeRef v) { return fn(T(v)); });
   }
+  static List create(const SourceRange& range, const std::vector<T>& subtrees) {
+    TreeList type_erased_sub {subtrees.begin(), subtrees.end()};
+    return List(Compound::create(TK_LIST, range, std::move(type_erased_sub)));
+  }
   size_t size() const {
     return tree_->trees().size();
   }
 };
 
 template <typename T>
-struct OptionView : public TreeView {
-  explicit OptionView(const TreeRef& tree) : TreeView(tree) {
+struct Maybe : public TreeView {
+  explicit Maybe(const TreeRef& tree) : TreeView(tree) {
     JIT_SCRIPT_ASSERT(tree, tree->kind() == TK_OPTION);
   }
+  /* implicit */ Maybe(const T& tree) : TreeView(tree) {}
   bool present() const {
     return tree_->trees().size() > 0;
   }
   T get() const {
-    JIT_SCRIPT_ASSERT(tree_, present());
-    return T(tree_->trees()[0]);
+    return T(tree_->trees().at(0));
   }
   TreeRef map(std::function<TreeRef(const T&)> fn) {
     return tree_->map([&](TreeRef v) { return fn(T(v)); });
   }
 };
 
 struct Ident : public TreeView {
-  // each subclass of TreeView provides:
-  // 1. a constructor that takes a TreeRef, and matches it to the right type.
   explicit Ident(const TreeRef& tree) : TreeView(tree) {
-    tree_->match(TK_IDENT, name_);
+    tree_->match(TK_IDENT);
   }
-  // 2. accessors that get underlying information out of the object
-  // in this case, we return the name of the identifier, and handle the
-  // converstion to a string in the method
   const std::string& name() const {
-    return name_->stringValue();
+    return subtree(0)->stringValue();
   }
-
-  // 3. a static method 'create' that creates the underlying TreeRef object
-  // for every TreeRef kind that has a TreeView, the parser always uses
-  // (e.g.) Ident::create rather than Compound::Create, this means that
-  // changes to the structure of Ident are always made right here rather
-  // than both in the parser and in this code
-  static TreeRef create(const SourceRange& range, const std::string& name) {
-    return Compound::create(TK_IDENT, range, {String::create(name)});
+  static Ident create(const SourceRange& range, const std::string& name) {
+    return Ident(Compound::create(TK_IDENT, range, {String::create(name)}));
   }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// Base types (production LHS)
+////////////////////////////////////////////////////////////////////////////////
+
+struct Type : public TreeView {
+  explicit Type(const TreeRef& tree) : TreeView(tree) {}
```

Comment:
Views normally check they are valid on construction, can you put an assert here that tree.kind is a valid type token?

Reply:
(we will probably need a set. It will be easy to maintain)